I am now seeing this error when running the UI e2e tests. It happens when running the same scheduling goal twice in a row.

```
2023-02-02T19:45:13.794313076Z.794313gov.nasa.jpl.aerie.scheduler.server.exceptions.ResultsProtocolFailure: The scheduler should not be deleting activity instances
gov.nasa.jpl.aerie.scheduler.server.exceptions.ResultsProtocolFailure: The scheduler should not be deleting activity instances
	at gov.nasa.jpl.aerie.scheduler.worker.services.SynchronousSchedulerAgent.storeFinalPlan(SynchronousSchedulerAgent.java:516)
	at gov.nasa.jpl.aerie.scheduler.worker.services.SynchronousSchedulerAgent.schedule(SynchronousSchedulerAgent.java:210)
	at gov.nasa.jpl.aerie.scheduler.worker.SchedulerWorkerAppDriver.main(SchedulerWorkerAppDriver.java:91)
Caused by: gov.nasa.jpl.aerie.scheduler.server.services.PlanServiceException: The scheduler should not be deleting activity instances
	at gov.nasa.jpl.aerie.scheduler.server.services.GraphQLMerlinService.updatePlanActivityDirectives(GraphQLMerlinService.java:405)
	at gov.nasa.jpl.aerie.scheduler.worker.services.SynchronousSchedulerAgent.storeFinalPlan(SynchronousSchedulerAgent.java:505)
	... 2 more
	```

--- a/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/model/ActivityInstance.java
+++ b/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/model/ActivityInstance.java
@@ -101,6 +101,24 @@ public record ActivityInstance(
                                 parameters, topParent);
   }
 
+  private static ActivityInstance of(SchedulingActivityInstanceId id, ActivityType type, Duration start, Duration duration, Map<String, SerializedValue> parameters, SchedulingActivityInstanceId topParent) {
+    return new ActivityInstance(id,
+                                type,
+                                start,
+                                duration,
+                                parameters,
+                                topParent);
+  }
+
+  public static ActivityInstance copyOf(ActivityInstance activityInstance, Duration duration){
+    return ActivityInstance.of(activityInstance.id,
+            activityInstance.type,
+            activityInstance.startTime,
+            duration,
+            new HashMap<>(activityInstance.arguments),
+            activityInstance.topParent);
+  }
+
   /**
    * create an activity instance based on the provided one (but adifferent id)
    *

--- a/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/simulation/SimulationFacade.java
+++ b/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/simulation/SimulationFacade.java
@@ -130,6 +130,28 @@ public class SimulationFacade {
     }
   }
 
+  /**
+   * Replaces an activity instance with another, strictly when they have the same id
+   * @param toBeReplaced the activity to be replaced
+   * @param replacement the replacement activity
+   */
+  public void replaceActivityFromSimulation(final ActivityInstance toBeReplaced, final ActivityInstance replacement){
+    if(toBeReplaced.type() != replacement.type()||
+       toBeReplaced.startTime() != replacement.startTime()||
+       !(toBeReplaced.arguments().equals(replacement.arguments()))) {
+      throw new IllegalArgumentException("When replacing an activity, you can only update the duration");
+    }
+    if(!insertedActivities.containsKey(toBeReplaced)){
+      throw new IllegalArgumentException("Trying to replace an activity that has not been previously simulated");
+    }
+    final var associated = insertedActivities.get(toBeReplaced);
+    insertedActivities.remove(toBeReplaced);
+    insertedActivities.put(replacement, associated);
+    final var simulationId = this.planActInstanceIdToSimulationActInstanceId.get(toBeReplaced.id());
+    this.planActInstanceIdToSimulationActInstanceId.remove(toBeReplaced.id());
+    this.planActInstanceIdToSimulationActInstanceId.put(replacement.id(), simulationId);
+  }
+
   public void simulateActivities(final Collection<ActivityInstance> activities)
   throws SimulationException {
     final var activitiesSortedByStartTime =

--- a/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java
+++ b/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java
@@ -188,6 +188,7 @@ public class PrioritySolver implements Solver {
       }
       final var allGeneratedActivities = simulationFacade.getAllGeneratedActivities(simulationFacade.getCurrentSimulationEndTime());
       processNewGeneratedActivities(allGeneratedActivities);
+      pullActivityDurationsIfNecessary();
     } else{
       //update simulation with regard to plan
       try {
@@ -223,7 +224,36 @@ public class PrioritySolver implements Solver {
       simulationFacade.simulateActivities(plan.getActivities());
       final var allGeneratedActivities = simulationFacade.getAllGeneratedActivities(problem.getPlanningHorizon().getEndAerie());
       processNewGeneratedActivities(allGeneratedActivities);
+      pullActivityDurationsIfNecessary();
     }
+  }
+
+  /**
+   * For activities that have a null duration (in an initial plan for example) and that have been simulated, we pull the duration and
+   * replace the original instance with a new instance that includes the duraiton, both in the plan and the simulation facade
+   */
+  public void pullActivityDurationsIfNecessary() {
+    final var toRemoveFromPlan = new ArrayList<ActivityInstance>();
+    final var toAddToPlan = new ArrayList<ActivityInstance>();
+    for (final var activity : plan.getActivities()) {
+      if (activity.duration() == null) {
+        final var duration = simulationFacade.getActivityDuration(activity);
+        if (duration.isPresent()) {
+          final var replacementAct = ActivityInstance.copyOf(
+              activity,
+              duration.get()
+              );
+          simulationFacade.replaceActivityFromSimulation(activity, replacementAct);
+          toAddToPlan.add(replacementAct);
+          toRemoveFromPlan.add(activity);
+          generatedActivityInstances = generatedActivityInstances.stream().map(pair -> pair.getLeft().equals(activity) ? Pair.of(replacementAct, pair.getRight()): pair).collect(Collectors.toList());
+          generatedActivityInstances = generatedActivityInstances.stream().map(pair -> pair.getRight().equals(activity) ? Pair.of(pair.getLeft(), replacementAct): pair).collect(Collectors.toList());
+        }
+      }
+    }
+
+    plan.remove(toRemoveFromPlan);
+    plan.add(toAddToPlan);
   }
 
   /**

--- a/scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/SchedulingIntegrationTests.java
+++ b/scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/SchedulingIntegrationTests.java
@@ -1348,6 +1348,35 @@ public class SchedulingIntegrationTests {
     assertEquals(96, results.updatedPlan().size());
   }
 
+  /**
+   * If you passed activities without duration to the scheduler in an initial plan, it would fail
+   */
+  @Test
+  public void testBugDurationInMicroseconds(){
+    final var results = runScheduler(
+        BANANANATION,
+        List.of(),
+        List.of(new SchedulingGoal(new GoalId(0L), """
+            export default (): Goal =>
+              Goal.ActivityRecurrenceGoal({
+                activityTemplate: ActivityTemplates.BakeBananaBread({ temperature: 325.0, tbSugar: 2, glutenFree: false }),
+                interval: Temporal.Duration.from({ hours: 12 }),
+              });
+            """, true)),
+        PLANNING_HORIZON);
+    final var results2 = runScheduler(
+        BANANANATION,
+        results.updatedPlan.stream().toList(),
+        List.of(new SchedulingGoal(new GoalId(0L), """
+            export default (): Goal =>
+              Goal.ActivityRecurrenceGoal({
+                activityTemplate: ActivityTemplates.BakeBananaBread({ temperature: 325.0, tbSugar: 2, glutenFree: false }),
+                interval: Temporal.Duration.from({ hours: 12 }),
+              });
+            """, true)),
+        PLANNING_HORIZON);
+    assertEquals(8, results.updatedPlan().size());
+  }
 
   @Test
   void test_inf_loop(){

